lib/decorateWithMetrics.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * Each key is a method name and the value is an array of calls metadata
+ * @type {Map<String, Array<CallMeta>>}
+ */
+const stats = new Map();
+
+/**
+ * Class instances are easily tracked by Dev tools memory snapshots
+ * Wrapping a result with such instance allows us to track it without
+ * incurring extra weight
+ */
+class TaggedItem {
+    constructor(data) {
+        this.ref = data;
+    }
+}
+
+/**
+ * Decorates an instance method with metrics capturing functionality
+ * The method needs to be async or to return a promise
+ * @param {Object} instance
+ * @param {String} methodName
+ */
+function decorateWithMetrics(instance, methodName) {
+    const originalMethod = instance[methodName];
+
+    if (stats.has(methodName)) {
+        throw new Error(`"${methodName}" is already decorated`);
+    }
+
+    stats.set(methodName, []);
+
+    // eslint-disable-next-line no-param-reassign
+    instance[methodName] = (...args) => {
+        const startTime = performance.now();
+        let result;
+        let error;
+
+        const promise = originalMethod.apply(instance, args);
+
+        // Then handlers added here are not affecting the returned promise as they aren't attached to it
+        promise.then(data => data && (result = new TaggedItem(data)))
```

Comment:
Would it be bad to create a `TaggedItem` even if `data` had no value? Just curious why we are preventing this since some `result` will be `undefined`.

Reply:
Some methods are `void` they don't return anything 
With `TaggedItem` we're trying to identify only actual data that uses memory

The check is a bit crude and should be updated to check only against `null` and `undefined` 

For example let say:
- you're in a chat with cleared stats (`resetMetrics` was just called)
- You switch to a different chat 
- You capture a snapshot in dev tools

The `TaggedItems` there would be pretty much how much data Onyx consumes for the 2nd chat - only Onyx data from storage, not rendered content and DOM stuff. We don't want to capture `TaggedItems` for void returns, though it would barely increase memory it would list more `TaggedItems` than the actual results returned from Onyx